### ISSUE TITLE
window program

### ISSUE THREAD

**Issue by mrGreenS** [olned/ssc2ce-python#8]
Hi!
When I launch the code
loop.run_until_complete(app.run_receiver())
the button in the window program stalls. But the "run_receiver" works fine and gives data to log. What is the way to avoid those stalls?

**Comment by olned** [comment]
Hi!
Could you give me a little more information about the environment in which you run the script? And which script do you run?

**Comment by mrGreenS** [comment]
Thanks for being online!
We are using Python 3.8, PyCharm community 2020.01 . We have taken  the code from the deribit_private.py file. Going to interact with PyQt5 form. Would like to start and stop subscription and send trading commands with pressing a button.

**Comment by olned** [comment]
It's my pleasure!
Do you use [asyncqt](https://github.com/gmarull/asyncqt) or something similar to get Qt event loop?

**Comment by olned** [comment]
I slightly changed  [the asyncqt example](https://github.com/gmarull/asyncqt/blob/master/examples/aiohttp_fetch.py). This example works for me. Please let me know if you solved the problem.

```
import sys
import asyncio

from asyncqt import QEventLoop, asyncSlot, asyncClose
from ssc2ce import Deribit
from PyQt5.QtWidgets import (
# from PySide2.QtWidgets import (
    QApplication, QWidget, QLabel, QLineEdit, QTextEdit, QPushButton,
    QVBoxLayout)


class MainWindow(QWidget):
    """Main window."""

    _DEF_URL = 'https://jsonplaceholder.typicode.com/todos/1'
    """str: Default URL."""

    _SESSION_TIMEOUT = 1.
    """float: Session timeout."""

    def __init__(self):
        super().__init__()

        self.setLayout(QVBoxLayout())

        self.lblStatus = QLabel('Idle', self)
        self.layout().addWidget(self.lblStatus)

        # self.editUrl = QLineEdit(self._DEF_URL, self)
        # self.layout().addWidget(self.editUrl)

        self.editResponse = QTextEdit('', self)
        self.layout().addWidget(self.editResponse)

        self.btnFetch = QPushButton('Fetch', self)
        self.btnFetch.clicked.connect(self.on_btnFetch_clicked)
        self.layout().addWidget(self.btnFetch)

        self.conn = Deribit()
        self.loop = asyncio.get_event_loop()
        self.feed_task = None
        self.conn.on_connect_ws = self.subscribe
        self.conn.method_routes += [("subscription", self.handle_subscription)]

        # self.session = aiohttp.ClientSession(
        #     loop=asyncio.get_event_loop(),
        #     timeout=aiohttp.ClientTimeout(total=self._SESSION_TIMEOUT))

    @asyncClose
    async def closeEvent(self, event):
        await self.session.close()

    @asyncSlot()
    async def on_btnFetch_clicked(self):
        self.btnFetch.setEnabled(False)
        self.lblStatus.setText('Fetching...')

        try:
            self.feed_task = asyncio.ensure_future(self.conn.run_receiver())
        except Exception as exc:
            self.lblStatus.setText(self.lblStatus.text() + '\nError: {}'.format(exc))
        finally:
            self.btnFetch.setEnabled(True)

    async def subscribe(self):
        await self.conn.send_public(request={
            "method": "public/subscribe",
            "params": {
                "channels": ["deribit_price_index.btc_usd"]
            }
        })

    async def handle_subscription(self, data):
        method = data.get("method")
        if method and method == "subscription":
            if data["params"]["channel"].startswith("deribit_price_index"):
                index_name = data["params"]["data"]["index_name"]
                price = data["params"]["data"]["price"]
                self.editResponse.append(f"Deribit Price Index {index_name.upper()}: {price}")


if __name__ == '__main__':
    app = QApplication(sys.argv)
    loop = QEventLoop(app)
    asyncio.set_event_loop(loop)

    mainWindow = MainWindow()
    mainWindow.show()

    with loop:
        sys.exit(loop.run_forever())
```

**Comment by mrGreenS** [comment]
That is the code I have been dreaming of for so long. Very useful! 
Appreciate it!
Thanks, man!

**Comment by olned** [comment]
You are welcome!

**Comment by olned** [comment]
You are welcome!
If you solved the problem, am I closing the issue?

**Comment by mrGreenS** [comment]
Could you explain why when I call `self.conn = Deribit()`
the following is not tranmitted client_id, client_secret, but the authorization still occurs.
In case I am trying to connect like this 
`self.conn = Deribit(client_id=client_id, client_secret=client_secret, auth_type=AuthType.CREDENTIALS, get_id=lambda: str(uuid4()))`
using false data everything is still working

**Comment by mrGreenS** [comment]
Could you please add to the mentioned above example of your code receiving the current position on the instrument and sending a buy order according some condition.

**Comment by olned** [comment]
> Could you explain why when I call `self.conn = Deribit()`
> the following is not tranmitted client_id, client_secret, but the authorization still occurs.
> In case I am trying to connect like this
> `self.conn = Deribit(client_id=client_id, client_secret=client_secret, auth_type=AuthType.CREDENTIALS, get_id=lambda: str(uuid4()))`
> using false data everything is still working

This should not work as you described. It’s hard for me to understand the reasons without seeing your code, but I suspect that it might have been the reason.

It can be related on the on_connect_ws handler that why I made small changes to class Deribit:
- If `auth_type` other than `AuthType.NONE` is passed to the constructor, `auth_login` is assigned to the `on_connect_ws` handler. Now you do not have to do it manually.
- The `on_response_error` handler specification has been changed, the `data: dict `parameter has been added. I think you just missed the error message about 

And I extended the deribit_private example. I created a class and add subscriptions which can be useful if you will analyze you orders and trades.

**Comment by olned** [comment]
> Could you please add to the mentioned above example of your code receiving the current position on the instrument and sending a buy order according some condition.

Yes I can do it, but it would be nice if you open a new issue and describe the example script in more detail. What event should be a trigger for placing an order, at what price, what size.

**Comment by mrGreenS** [comment]
> I slightly changed [the asyncqt example](https://github.com/gmarull/asyncqt/blob/master/examples/aiohttp_fetch.py). This example works for me. Please let me know if you solved the problem.
> 
> ```
> import sys
> import asyncio
> 
> from asyncqt import QEventLoop, asyncSlot, asyncClose
> from ssc2ce import Deribit
> from PyQt5.QtWidgets import (
> # from PySide2.QtWidgets import (
>     QApplication, QWidget, QLabel, QLineEdit, QTextEdit, QPushButton,
>     QVBoxLayout)
> 
> 
> class MainWindow(QWidget):
>     """Main window."""
> 
>     _DEF_URL = 'https://jsonplaceholder.typicode.com/todos/1'
>     """str: Default URL."""
> 
>     _SESSION_TIMEOUT = 1.
>     """float: Session timeout."""
> 
>     def __init__(self):
>         super().__init__()
> 
>         self.setLayout(QVBoxLayout())
> 
>         self.lblStatus = QLabel('Idle', self)
>         self.layout().addWidget(self.lblStatus)
> 
>         # self.editUrl = QLineEdit(self._DEF_URL, self)
>         # self.layout().addWidget(self.editUrl)
> 
>         self.editResponse = QTextEdit('', self)
>         self.layout().addWidget(self.editResponse)
> 
>         self.btnFetch = QPushButton('Fetch', self)
>         self.btnFetch.clicked.connect(self.on_btnFetch_clicked)
>         self.layout().addWidget(self.btnFetch)
> 
>         self.conn = Deribit()
>         self.loop = asyncio.get_event_loop()
>         self.feed_task = None
>         self.conn.on_connect_ws = self.subscribe
>         self.conn.method_routes += [("subscription", self.handle_subscription)]
> 
>         # self.session = aiohttp.ClientSession(
>         #     loop=asyncio.get_event_loop(),
>         #     timeout=aiohttp.ClientTimeout(total=self._SESSION_TIMEOUT))
> 
>     @asyncClose
>     async def closeEvent(self, event):
>         await self.session.close()
> 
>     @asyncSlot()
>     async def on_btnFetch_clicked(self):
>         self.btnFetch.setEnabled(False)
>         self.lblStatus.setText('Fetching...')
> 
>         try:
>             self.feed_task = asyncio.ensure_future(self.conn.run_receiver())
>         except Exception as exc:
>             self.lblStatus.setText(self.lblStatus.text() + '\nError: {}'.format(exc))
>         finally:
>             self.btnFetch.setEnabled(True)
> 
>     async def subscribe(self):
>         await self.conn.send_public(request={
>             "method": "public/subscribe",
>             "params": {
>                 "channels": ["deribit_price_index.btc_usd"]
>             }
>         })
> 
>     async def handle_subscription(self, data):
>         method = data.get("method")
>         if method and method == "subscription":
>             if data["params"]["channel"].startswith("deribit_price_index"):
>                 index_name = data["params"]["data"]["index_name"]
>                 price = data["params"]["data"]["price"]
>                 self.editResponse.append(f"Deribit Price Index {index_name.upper()}: {price}")
> 
> 
> if __name__ == '__main__':
>     app = QApplication(sys.argv)
>     loop = QEventLoop(app)
>     asyncio.set_event_loop(loop)
> 
>     mainWindow = MainWindow()
>     mainWindow.show()
> 
>     with loop:
>         sys.exit(loop.run_forever())
> ```
I mean that the script is running without  assigning  client_id,  client_secret  and it still establishes connection to Deribit.

**Comment by olned** [comment]
Deribit provides the ability to make some subscriptions without authentication. Therefore, you can connect to the server and [subscribe to some public channels ](https://docs.deribit.com/#public-subscribe). In the example, `/public/subscribe` is  used.

```
    async def subscribe(self):
        await self.conn.send_public(request={
            "method": "public/subscribe",
            "params": {
                "channels": ["deribit_price_index.btc_usd"]
            }
        })
```

**Comment by mrGreenS** [comment]
> I slightly changed [the asyncqt example](https://github.com/gmarull/asyncqt/blob/master/examples/aiohttp_fetch.py). This example works for me. Please let me know if you solved the problem.
> 
> ```
> import sys
> import asyncio
> 
> from asyncqt import QEventLoop, asyncSlot, asyncClose
> from ssc2ce import Deribit
> from PyQt5.QtWidgets import (
> # from PySide2.QtWidgets import (
>     QApplication, QWidget, QLabel, QLineEdit, QTextEdit, QPushButton,
>     QVBoxLayout)
> 
> 
> class MainWindow(QWidget):
>     """Main window."""
> 
>     _DEF_URL = 'https://jsonplaceholder.typicode.com/todos/1'
>     """str: Default URL."""
> 
>     _SESSION_TIMEOUT = 1.
>     """float: Session timeout."""
> 
>     def __init__(self):
>         super().__init__()
> 
>         self.setLayout(QVBoxLayout())
> 
>         self.lblStatus = QLabel('Idle', self)
>         self.layout().addWidget(self.lblStatus)
> 
>         # self.editUrl = QLineEdit(self._DEF_URL, self)
>         # self.layout().addWidget(self.editUrl)
> 
>         self.editResponse = QTextEdit('', self)
>         self.layout().addWidget(self.editResponse)
> 
>         self.btnFetch = QPushButton('Fetch', self)
>         self.btnFetch.clicked.connect(self.on_btnFetch_clicked)
>         self.layout().addWidget(self.btnFetch)
> 
>         self.conn = Deribit()
>         self.loop = asyncio.get_event_loop()
>         self.feed_task = None
>         self.conn.on_connect_ws = self.subscribe
>         self.conn.method_routes += [("subscription", self.handle_subscription)]
> 
>         # self.session = aiohttp.ClientSession(
>         #     loop=asyncio.get_event_loop(),
>         #     timeout=aiohttp.ClientTimeout(total=self._SESSION_TIMEOUT))
> 
>     @asyncClose
>     async def closeEvent(self, event):
>         await self.session.close()
> 
>     @asyncSlot()
>     async def on_btnFetch_clicked(self):
>         self.btnFetch.setEnabled(False)
>         self.lblStatus.setText('Fetching...')
> 
>         try:
>             self.feed_task = asyncio.ensure_future(self.conn.run_receiver())
>         except Exception as exc:
>             self.lblStatus.setText(self.lblStatus.text() + '\nError: {}'.format(exc))
>         finally:
>             self.btnFetch.setEnabled(True)
> 
>     async def subscribe(self):
>         await self.conn.send_public(request={
>             "method": "public/subscribe",
>             "params": {
>                 "channels": ["deribit_price_index.btc_usd"]
>             }
>         })
> 
>     async def handle_subscription(self, data):
>         method = data.get("method")
>         if method and method == "subscription":
>             if data["params"]["channel"].startswith("deribit_price_index"):
>                 index_name = data["params"]["data"]["index_name"]
>                 price = data["params"]["data"]["price"]
>                 self.editResponse.append(f"Deribit Price Index {index_name.upper()}: {price}")
> 
> 
> if __name__ == '__main__':
>     app = QApplication(sys.argv)
>     loop = QEventLoop(app)
>     asyncio.set_event_loop(loop)
> 
>     mainWindow = MainWindow()
>     mainWindow.show()
> 
>     with loop:
>         sys.exit(loop.run_forever())
> ```

When using ssc2ce version 0.12.1  this code generates the following mistake 

D:\Python\Python38\lib\site-packages\ssc2ce\deribit\deribit.py:457: RuntimeWarning: coroutine 'MainWindow.handle_subscription' was never awaited
  handler(data)
RuntimeWarning: Enable tracemalloc to get the object allocation traceback

**Comment by olned** [comment]
Fixed in v0.12.2
Please check if it works now
